Grindstone event

ferriarnus opened the pull request:
An event implementation for the grindstone. This is heavily based on the anvil event. I've made this implementation based on the talks in the discord chat, so any comments/remarks are certainly welcome if I've missed something.

A quick question: similar to the anvil events, these events do not handle changes in the input items. Would another event (or a "post" type event) be good to achieve this? Currently the vanilla behavior is to set the slots to empty, deleting whatever was in them, even if it was a full stack.

ferriarnus committed:
Grindstone event

An event implementation for the grindstone. This is heavily based on the anvil event. I've made this implementation based on the talks in the discord chat, so any comments/remarks are certainly welcome if I've missed something.

XFactHD commented on src/main/java/net/minecraftforge/common/ForgeHooks.java:
```diff
@@ -674,6 +677,24 @@ public static float onAnvilRepair(Player player, @Nonnull ItemStack output, @Non
+        }
+
+        outputSlot.setItem(0, e.getOutput());
+        xp = e.getXp();
```
All assignments to the `xp` parameter are dropped when this method returns, they will never reach the `GrindstoneMenu#xp` variable you added.

ferriarnus replied:
Oh that's stupid of me. The simpelst way to fix this is make use of an int[] or some "mutable int" instead. What would be the best way?

XFactHD replied:
I would say `int[]` or the `MutableInt` from apache-commons.

diesieben07 replied:
Can you not return the int instead and use two negative numbers to represent true and false?

diesieben07 replied:
Can you not return the int instead and use two negative numbers to represent true and false?

ferriarnus replied:
Of course that also works! I'm really not paying attention.

KnightMiner commented on patches/minecraft/net/minecraft/world/inventory/GrindstoneMenu.java.patch:
```diff
@@ -1,5 +1,45 @@
+       this.m_38897_(new Slot(this.f_39560_, 0, 49, 19) {
+          public boolean m_5857_(ItemStack p_39607_) {
+-            return p_39607_.m_41763_() || p_39607_.m_150930_(Items.f_42690_) || p_39607_.m_41793_();
++            return true;
```
This may cause a desync between a forge server and a vanilla client or vice versa, have you tested that case? Should be able to launch both through the dev environment, use offline mode for the server.

ferriarnus replied:
I've checked forge server vanilla client, the items would go in the grindstone no problem. I've tried to join a vanilla world from lan using forge, but that did not work (I couldn't join). I'll need to setup a vanilla server to test this?

SizableShrimp replied:
If you remove the --uuid 0 arg from the forge run config then you should be able to join LAN stuff

ferriarnus replied:
Thanks for the help. I don't see any obvious issues. When forge is the server any item can go in the slot no issue. If vanilla is the server the item won't go in (if you press really fast it will go in for a fraction of a second, but it's hard to notice).

KnightMiner commented on src/main/java/net/minecraftforge/event/GrindstoneUpdateEvent.java:
```diff
@@ -0,0 +1,95 @@
+{
+	
+    private final ItemStack top;
+    private final ItemStack bottem;
```
typo, should be `bottom`

ferriarnus committed:
small changes

Made use of Integer.MIN_VALUE as a return value for empty output event.

ferriarnus committed:
second event

Made a second event, to change how the input items are changed after the result item is taken out.

ghost commented:
I just wanted to throw my hat into the ring and strongly encourage the forge team to add these events into the game, mods should be able to control what happens to their content in vanilla containers.

TheCurle commented on src/test/java/net/minecraftforge/debug/GrindstoneEventTest.java:
```diff
@@ -0,0 +1,59 @@
+package net.minecraftforge.debug;
```
License, please.

Matyrobbrt commented:
This has been effectively superseded by #9090.
